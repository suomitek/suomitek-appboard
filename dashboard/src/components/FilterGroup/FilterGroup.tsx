import React, { useState } from "react";

import MultiCheckbox from "components/js/MultiCheckbox";

interface IFilterGroupProps {
  name: string;
  options: string[];
  onChange: (newValue: string[]) => void;
}

function FilterGroup({ name, options, onChange: propsOnChange }: IFilterGroupProps) {
  const [value, setValue] = useState([] as string[]);
  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { value: targetValue } = e.target;
    const index = value.indexOf(targetValue);
    let newValue = [];
    if (index > -1) {
      // Already checked, remove
      value.splice(index, 1);
      newValue = [...value];
    } else {
      // Not checked, add
      newValue = value.concat(targetValue);
    }
    setValue(newValue);
    propsOnChange(newValue);
  };
  return <MultiCheckbox name={name} options={options} span={1} value={value} onChange={onChange} />;
}

export default FilterGroup;
