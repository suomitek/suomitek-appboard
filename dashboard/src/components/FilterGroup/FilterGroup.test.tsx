import * as React from "react";

import MultiCheckbox from "components/js/MultiCheckbox";
import { act } from "react-dom/test-utils";
import { defaultStore, mountWrapper } from "shared/specs/mountWrapper";
import FilterGroup from "./FilterGroup";

const defaultProps = {
  name: "test",
  options: ["foo", "bar"],
  onChange: jest.fn(),
};

it("renders a multicheckbox", () => {
  const wrapper = mountWrapper(defaultStore, <FilterGroup {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});

it("calls onChange function", () => {
  const onChange = jest.fn();
  const wrapper = mountWrapper(defaultStore, <FilterGroup {...defaultProps} onChange={onChange} />);
  act(() => {
    wrapper.find(MultiCheckbox).prop("onChange")({ target: { value: "foo" } });
  });
  expect(onChange).toHaveBeenCalledWith(["foo"]);
  // Force re-render
  wrapper.setProps({ ...defaultProps, onChange });
  // Adds a new item to the filter
  act(() => {
    wrapper.find(MultiCheckbox).prop("onChange")({ target: { value: "bar" } });
  });
  expect(onChange).toHaveBeenCalledWith(["foo", "bar"]);
  // Force re-render
  wrapper.setProps({ ...defaultProps, onChange });
  // Removes an item
  act(() => {
    wrapper.find(MultiCheckbox).prop("onChange")({ target: { value: "foo" } });
  });
  expect(onChange).toHaveBeenCalledWith(["bar"]);
});
