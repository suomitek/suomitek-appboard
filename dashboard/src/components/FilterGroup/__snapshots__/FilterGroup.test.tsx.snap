// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a multicheckbox 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <FilterGroup
        name="test"
        onChange={[MockFunction]}
        options={
          Array [
            "foo",
            "bar",
          ]
        }
      >
        <MultiCheckbox
          name="test"
          onChange={[Function]}
          options={
            Array [
              "foo",
              "bar",
            ]
          }
          span={1}
          value={Array []}
        >
          <div
            className="multicheckbox-wrapper"
            style={
              Object {
                "--col-lg-num": 1,
                "--col-md-num": 1,
                "--col-sm-num": 1,
                "--col-xl-num": 1,
              }
            }
          >
            <div
              className="clr-checkbox-wrapper"
              key="0"
            >
              <input
                checked={false}
                id="test-0"
                name="test"
                onChange={[Function]}
                type="checkbox"
                value="foo"
              />
              <label
                className="clr-control-label"
                htmlFor="test-0"
              >
                foo
              </label>
            </div>
            <div
              className="clr-checkbox-wrapper"
              key="1"
            >
              <input
                checked={false}
                id="test-1"
                name="test"
                onChange={[Function]}
                type="checkbox"
                value="bar"
              />
              <label
                className="clr-control-label"
                htmlFor="test-1"
              >
                bar
              </label>
            </div>
          </div>
        </MultiCheckbox>
      </FilterGroup>
    </Router>
  </BrowserRouter>
</Provider>
`;
